src/gt4py/analysis/passes.py

```diff
@@ -1290,6 +1290,89 @@ def apply(cls, transform_data: TransformData) -> None:
         cls.DemoteSymbols.apply(transform_data.implementation_ir, demotables)
 
 
+class ReduceTemporaryStoragesPass(TransformPass):
+    """Demote 3D temporaries only used within a single stage to 2D temporaries."""
+
+    class ReducibleFieldsCollector(gt_ir.IRNodeVisitor):
+        @classmethod
+        def apply(cls, node: gt_ir.StencilImplementation) -> Set[str]:
+            collector = cls()
+            return collector(node)
+
+        def __call__(self, node: gt_ir.StencilImplementation) -> Set[str]:
+            assert isinstance(node, gt_ir.StencilImplementation)
+            self.interval: gt_ir.AxisInterval = None
+            self.iteration_order: gt_ir.IterationOrder = None
+            self.reduced_fields: Dict[str, gt_ir.AxisInterval] = {
+                temp_field: None for temp_field in node.temporary_fields
+            }
+            self.visit(node)
+            return set(self.reduced_fields.keys())
+
+        def visit_MultiStage(self, node: gt_ir.MultiStage) -> None:
+            self.iteration_order = node.iteration_order
+            self.generic_visit(node)
+
+        def visit_ApplyBlock(self, node: gt_ir.ApplyBlock) -> None:
+            self.interval = node.interval
+            self.generic_visit(node)
+
+        def visit_FieldRef(self, node: gt_ir.FieldRef) -> None:
+            field_name = node.name
+            if field_name in self.reduced_fields:
+                if self.iteration_order == gt_ir.IterationOrder.PARALLEL:
+                    # Do not reduce fields accessed from parallel k-intervals
+                    self.reduced_fields.pop(field_name)
+                else:
+                    offsets: List[int] = list(node.offset.values())
+                    if offsets[-1] != 0:
+                        # Do not reduce fields with k-offsets
+                        self.reduced_fields.pop(field_name)
+                    else:
+                        # Do not reduce fields that are accessed across different k-intervals
+                        interval = self.reduced_fields[field_name]
+                        if interval is not None and interval != self.interval:
+                            self.reduced_fields.pop(field_name)
+                        else:
+                            self.reduced_fields[field_name] = self.interval
```

Comment:
If a temporary would be written in one stage, and then read in another with the same interval, this would still be considered as reducible, although it shouldn't be. I think instead of the interval it should be something that  identifies the ApplyBlock here.  (e.g. the id of the apply block) Or, alternatively add some logic at the visit_ApplyBlock level to filter out ones that have been used before.

Reply:
This is an excellent point. If horizontal offsets require new stages and therefore apply blocks, would this pass even be useful in that case? Would not the `DemoteTemporaries` pass have already demoted them to scalars?

Reply:
Then consider there being a statement that has to be separate for whatever other reason in between, that doesn't access said field at all. You'd end up with a similar situation. 

Either way, these are optimization passes and in my opinion, they should rely on the assumption that another optimization pass has run first. They should be written s.t. it will be safe to turn them off independently, or reorder etc, without violating validity.

Reply:
I completely agree that compiler passes should be independent of one another and produce valid output IR from valid input IR. I am merely curious from a practical perspective, whether there is a case in which a temporary field can be demoted to `IJ` but not to a scalar...